### ISSUE TITLE
Scope of the language / deviations from Python

### ISSUE THREAD

**Issue by apsz3** [blueloveTH/pocketpy#18]
I would like to get a better sense of how PocketPy intends to develop as a language. Does it aim to strictly retain as close a relationship to Python as possible, in terms of syntax and functionality? Or, does it aim to simply be Python-like in spirit?

For example, here are some non-Python features / syntax I would like to see (and will probably try to implement on my fork):

* Dotted access for dictionary keys

```python
d = {"a": 1} ; assert d.a == 1
```

* Less verbose `lambda` syntax. One idea:

```python
add = x, y -> x + y
const = None -> 1
((x, y, z) -> x + y + z)(1,1,1)
```

* Labeled `for` loops. One idea:

```
for x in Iter1 as outer:
  for y in Iter2 as inner:
    if f(x, y):
     break outer # Break out of both loops 
```

Another thing I noticed is that `dict` is implemented in Pocketpy, instead of in CPP. Could you explain the motivation for this? Is this motivated by keeping the language core as small as possible? 

Generally, I am trying to gauge what sort of contributions that extend the language, would be welcome.

**Comment by apsz3** [comment]
Another instance where I was not sure what the best approach was: I was going to modify `isinstance` to have better error handling if `cls` is not a type. (Right now it is an `assert` when it probably should be a `TypeError`.) CPython allows `cls` to be also a tuple of types. 

In Pocketpy, would we want to implement the error handling permissively, or strictly? E.g., would we iterate over `cls` to see if each item is a valid type, before proceeding with the rest of the function? Or, would we not check for invalid args initially, i.e. just iterate over it, and `raise` only if one of the iterands was not a valid type? The tradeoff is performance, vs  enhanced error handling / limiting side-effects. 

Speaking for my own desires, what I want out of a scripting language I have in mind, is above anything else avoiding footguns. Since I would be using this for gameplay scripting (as I believe you are), I want to prevent subtle errors as much as possible.

**Comment by blueloveTH** [comment]
In my vision, PocketPy is 90% python+10% extended features. We care about compability to CPython. If we can achieve it, "some" existing CPython based package can be reused/migrated. When adding new features, they must not conflict with python iteself.

Compatibility is not strict. The princple is, **let normal/obvious python code is compatible.** For example, 1-tuple or empty tuple is not of that.

**Comment by blueloveTH** [comment]
Yes, `dict` is impl in PocketPy itself. I name this "self-interpreted". It makes the core small and less bugs. Because when we do things in python, it tests the core in C and proves its correctness.

**Comment by blueloveTH** [comment]
However, currently we focus on CPython compatible features and try to make them stable. For instance, `setcomp`, `dictcomp`, `asyncio`, `nested function`, `decorator`, `static class variables` and overall performance...

Syntax sugars like your new lambda forms are not at the highest priority.

**Comment by blueloveTH** [comment]
error handling is permissively. use assert is ok and clear.

**Comment by apsz3** [comment]
> Some special instructions could be made to alleviate performance loss when impl builtins classes.

Yeah, this was my main concern.

Overall, thank you for the clarifications, this is helpful. 

Should we keep this issue open for further discussion / questions about the language scope / roadmap?
